### PR TITLE
[C] Add support for new preprocessor directives

### PR DESCRIPTION
**Pull request by braewoods**
This adds support for the new #elifdef and #elifndef preprocessor directives that are a combination of #else with #ifdef or #ifndef.

C++, Objective-C, and Objective-C++ are also updated because they derive from the C syntax family as well.

This has already made it into recent C and C++ compilers and is part of the C23 and C++23 standards.

### CODE REVIEW

**Comment by braewoods** [comment]
Ok. I thought I'd found something I needed to adjust but turns out it wasn't the case.

**Comment by michaelblyons** [comment]
Can you add some assertions in the test modification?

**Comment by braewoods** [comment]
> Can you add some assertions in the test modification?

All I did was extend the existing rules that already handle `elif` and `else` directives to include the new ones, and the only test case I could find for `elif`. What assertions could I even add that would be useful?

File: C++/syntax_test_c.c [hunk 1]
```diff
@@ -437,8 +437,16 @@ if (1) {
 if (2) {
 # elif BAZ
 if (3) {
-# else
+#elifdef BAR2
```
Comment (deathaxe) [inline C++/syntax_test_c.c:440]: What Michael is asking for is adding some test assertions.

Example

```suggestion
#elifdef BAR2
/* <- keyword.control.import */
/*^^^^^^ keyword.control.import */
```

The choosen section however seems to rather be for testing correct handling of multiple `{` with intermediate directives.

Tests targeting `#ifdef` seem to be located at https://github.com/sublimehq/Packages/blob/493cb8218a2d77128fb54ebf096256aa5cdeb019/C%2B%2B/syntax_test_c.c#L583-L584

File: C++/C.sublime-syntax [hunk 2]
```diff
@@ -1041,8 +1041,8 @@ contexts:
       captures:
         1: meta.preprocessor.c keyword.control.import.c
       pop: true
-    - match: (?=^\s*#\s*(elif|else)\b)
-      push: preprocessor-elif-else-branch-global
+    - match: (?=^\s*#\s*(elif|elifdef|elifndef|else)\b)
+      push: preprocessor-elif-elifdef-elifndef-else-branch-global
```
Comment (deathaxe) [inline C++/C.sublime-syntax:1045]: We probably don't need to rename contexts. Adding keywords to patterns is sufficient.

Comment (braewoods) [reply C++/C.sublime-syntax]: Reverted in latest revision.

**Comment by braewoods** [comment]
I've attempted to revise it given the provided feedback. Is there anything else?

File: C++/syntax_test_c.c [hunk 3]
```diff
@@ -589,9 +589,17 @@ int foo(int val, float val2[])
     if (val == -1) {
 /*  ^^ keyword.control */
 /*                 ^ meta.block meta.block punctuation.section.block.begin */
+#elifdef BAR
+ /* ^ meta.block meta.block */
```
Comment (deathaxe) [inline C++/syntax_test_c.c:593]: We probably want to make sure, those new keywords are scoped `keyword` as this is the major change of this PR.

```suggestion
 /* ^ meta.block meta.block keyword.control.import */
```

**Review by deathaxe — COMMENTED** [review]
A little note/suggestion about syntax tests, but otherwise LGTM!

File: C++/syntax_test_c.c (hunk 3 above)
Comment (braewoods) [reply C++/syntax_test_c.c:593]: > We probably want to make sure, those new keywords are scoped `keyword` as this is the major change of this PR.

Well that's the strange thing. When I tried to do that, it failed the assertion when I ran the syntax tests from within Sublime Text 4180. `keyword.control.import` would work for `#elif` but none of my new additions and I can't determine why.

Comment (deathaxe) [reply C++/syntax_test_c.c]: Well, checked out your branch and all scopes are applied as expected. Looks all good.

Have you symlinked C++ and Objective-C++ folders into your ST's _Packages_ directory, so ST can see the changes?

To not bother you with details, I've updated tests for you.

**Review by deathaxe — APPROVED** [review]

File: C++/syntax_test_c.c (hunk 3 above)
Comment (braewoods) [reply C++/syntax_test_c.c]: I re-ran the syntax tests from Sublime Text and got these from your recent changes:
```Packages/User/C++/syntax_test_c.c:592:1
error: scope does not match
592 | #elifdef BAR
593 |  /* <- keyword.control.import */
    |  ^ this location did not match
actual:
    | ^^^^^^^^^ source.c meta.function.c meta.block.c meta.block.c 

Packages/User/C++/syntax_test_c.c:596:1
error: scope does not match
596 | #elifndef BAZ
597 |  /* <- keyword.control.import */
    |  ^ this location did not match
actual:
    | ^^^^^^^^^^ source.c meta.function.c meta.block.c meta.block.c meta.block.c 

FAILED: 2 of 674 assertions in 1 files failed
[Finished]```

Comment (braewoods) [reply C++/syntax_test_c.c]: I see. So that's the issue. I thought it was rather strange. It's working now.

**Review by michaelblyons — APPROVED** [review]

**Pull request merged**